File: card_layout/internal/domain/player/player.go
package player

import (
	"card.layout/internal/domain"
)

type player struct {
	name string
	hand *domain.Hand
	game *domain.Game
}

func (p *player) GetName() string {
	return p.name
}

func (p *player) GetHand() *domain.Hand {
	return p.hand
}

func (p *player) DrawCard(deck *domain.Deck) {
	p.hand.AddCard(deck.DrawCard())
}

func (p *player) HandSize() int {
	return p.hand.Size()
}

type Strategy interface {
	NameHimself() string
	PickCard(orders []int, hand *domain.Hand) domain.Card
}

File: 1_1_cpp/domain/player/player.cpp
#include "player.h"

PlayerImpl::PlayerImpl(Showdown* s): point(0), hand(new Hand()), exchangeAble(true), showdown(s), exchangeHandData(new exchangeHand()) {}


PlayerName PlayerImpl::GetName() const {
    return name;
}

int PlayerImpl::GetPoint() const  {
    return point;
}
void PlayerImpl::GainPoint()  {
    point++;
}
Hand* PlayerImpl::GetHands()  {
    return hand;
}
void PlayerImpl::SetHands(Hand* h)  {
    hand = h;
}

void PlayerImpl::ExchangeHand(IPlayer* exchangee) {
    Hand* p1Hand = GetHands();
    Hand* p2Hand = exchangee->GetHands();

    SetHands(p2Hand);
    exchangee->SetHands(p1Hand);
}

void PlayerImpl::DrawCard(Deck* deck) {
    hand->AddCard(deck->DrawCard());
}

void PlayerImpl::TakeTurn1(IPlayer* player) {
    player->MakeExchangeHandDecision();
}

void PlayerImpl::TakeTurn2(IPlayer* player) {
    player->PickCard();
}